libmscore/fret.cpp

```diff
@@ -479,49 +663,355 @@ void FretDiagram::read(XmlReader& e)
+      {
+      FretItem::Barre b = barre(fret);
+
+      if (!b.exists()) {
+            if (string < _strings - 1) {
+                  _barres[fret] = FretItem::Barre(string, -1);
+                  removeDotsMarkers(string, -1, fret);
+                  setDot(string, fret);
+                  }
+            } 
+      else if (b.endString == -1 && b.startString < string) {
+            _barres[fret].endString = string;
+            setDot(string, fret);
+            }
+      else {
+            removeDotsMarkers(b.startString, b.endString, fret);
+            removeBarre(fret);
+            }
+      }
+
+//---------------------------------------------------------
+//   undoSetFretDot
+//---------------------------------------------------------
+
+void FretDiagram::undoSetFretDot(int _string, int _fret, bool _add /*= true*/, FretDotType _dtype /*= FretDotType::NORMAl*/)
+      {
+      for (ScoreElement* e : linkList()) {
+            FretDiagram* fd = toFretDiagram(e);
+            fd->score()->undo(new FretDot(fd, _string, _fret, _add, _dtype));
+            }
+      }
+
+//---------------------------------------------------------
+//   undoSetFretMarker
+//---------------------------------------------------------
+
+void FretDiagram::undoSetFretMarker(int _string, FretMarkerType _mtype)
+      {
+      for (ScoreElement* e : linkList()) {
+            FretDiagram* fd = toFretDiagram(e);
+            fd->score()->undo(new FretMarker(fd, _string, _mtype));
+            }
+      }
+
+//---------------------------------------------------------
+//   undoSetFretBarre
+//---------------------------------------------------------
+
+void FretDiagram::undoSetFretBarre(int _string, int _fret)
+      {
+      for (ScoreElement* e : linkList()) {
+            FretDiagram* fd = toFretDiagram(e);
+            fd->score()->undo(new FretBarre(fd, _string, _fret));
+            }
+      }
+
+//---------------------------------------------------------
+//   removeBarre
+//    Remove a barre on a given fret.
+//---------------------------------------------------------
+
+void FretDiagram::removeBarre(int f)
+      {
+      auto it = _barres.find(f);
+      _barres.erase(it);
+      }
+
+//---------------------------------------------------------
+//   removeBarres
+//    Remove barres crossing a certain point. Fret of 0 means any point along
+//    the string.
+//---------------------------------------------------------
+
+void FretDiagram::removeBarres(int string, int fret /*= 0*/)
+      {
+      auto iter = _barres.begin();
+      while (iter != _barres.end()) {
+            int bfret = iter->first;
+            FretItem::Barre b = iter->second;
+
+            if (b.exists() && b.startString <= string && (b.endString >= string || b.endString == -1)) {
+                  if (fret > 0 && fret != bfret)
+                        ++iter;
+                  else
+                        iter = _barres.erase(iter);
+            }
```

Comment:
This brace should be indented 6 more spaces.

Reply:
Fixed in next commit.